[uTVM] Initial BYOC support with c-source module

manupa-arm opened the pull request:
* Current flow uses saves the module as a c source and compiles it seperately. This blocks multi-module compilation.
* This patch upgrades it to use export_library which supports external modules.
* Added BYOC support for uTVM and testing
* A new member for IRModule to carry external function names as they are required to create the function registry when creating a system-lib.

manupa-arm commented:
cc: @tqchen @areusch @zhiics @comaniac

comaniac commented:
I'm not familiar with uTVM. Out of curiosity, how uTVM uses BYOC codegen C? I ask because BYOC codegen C is just for demonstration purpose so its functionality is not complete.

manupa-arm commented:
@comaniac , Likewise this is also for demonstration purpose but in the uTVM contex. i.e., if we can annotate and offload a set of operations and produce a c-source that can do the required computations and return the DLTensors, this is to show it works in the uTVM context. 

Thus, we do not intend to use "ccompiler" as the solution rather to show a external codegen that could create c-source the place of the "ccompiler", the changes done here should enable the compilation of a such a source. Moreover, we do not prefer the metadata Packing of Imports (either using PackImports* methods) as they will re-create the artifacts in the volatile memory. However, having a c-source gives much better control as to where the artifacts should be placed through a standard linker script.

comaniac commented on src/relay/backend/contrib/codegen_c/codegen_c.h:
```diff
@@ -235,14 +289,14 @@ class CodegenCBase {
       }
       this->PrintIndents();
-      code_stream_ << "std::memcpy(out" << i << ", " << outs[i].name << ", 4 * " << outs[i].size
+      code_stream_ << "memcpy(out" << i << ", " << outs[i].name << ", 4 * " << outs[i].size
```
Did we use the std namespace somewhere?

manupa-arm replied:
I dont think we need the namespace. Do we ?

areusch replied:
I think this explicitly avoids use of compiler builtins, which are sometimes perhaps broken?

comaniac reviewed (commented):
LGTM to the codegen C part.

areusch commented on src/relay/backend/contrib/codegen_c/codegen.cc:
```diff
@@ -215,28 +214,37 @@ class CodegenC : public MemoizedExprTranslator<std::vector<Output>>, public Code
+    code_stream_ << "#include <string.h>\n";
+    if (!variables.empty()) {
+      // These are only needed to handle metadata copying
+      code_stream_ << "#include <tvm/runtime/container.h>\n";
```
just wondering--I know this external compiler uses c++ but I don't think we should require it for µTVM. it's fine to be a demo, though.

manupa-arm replied:
Yea, this part is not expected to run with uTVM as it involves copying Array of NDArrays which is something we dont want in the context of uTVM, hence the gating as this is a demo.

manupa-arm commented on python/tvm/micro/build.py:
```diff
@@ -162,7 +158,12 @@ def build_static_runtime(
         libs.append(compiler.library(lib_build_dir, lib_srcs, lib_opts))
 
-    libs.append(compiler.library(mod_build_dir, [mod_src_path], generated_lib_opts))
+    libs.append(
```
No, the export_library will create a temporary directory to save source files and it will not live beyond the scope of the function. So it did not occur to me saving the sources is a feature (I thought it was a side-effect of this flow). Thus, I ll just say what I generally do: use pdb and do get_source to see the c-source created. Anyhow, do we think this would be a user-facing feature ?

if so we can do two things : 
A1 : Traverse module hierarchy and dump the sources here independently of export library, here.
A2 :  More generally  we can introduce an attr to target_kind "c" in the lines of something like "-dbgdir=<dbgdir>" and make the export_library to use that directory the dump the intermediary sources.

I think A2 would be more generic, but would be beyond the scope of this PR. 
Thus if people agree, I could just do A1 here for now with a TODO for A2 at some point. 
What do you think ?

areusch replied:
I think in the future we should consider generating full projects (I.e. zephyr projects) at which point we should have a defined way to describe module and library sources to the generators. saving generated code to the workspace tree means that (if the Workspace is operating with debug=True, you get source included in your debugger. that can come in handy when the thing causing a CHECK-fail is memory overflow from a TVMBackendAllocWorkspace inside an operator function. So I think we want A1 but we also want to compile the library from the user-accessible source files.

manupa-arm replied:
In order to satisfy that requirement, we could still make the export_library create the artifacts in a user given workspace directory. Thus, we would not have this split flow of exporting the IRModules in tvm vs utvm and thats what I meant by A2. 

On other hand of generating projects, I think we should discuss it in a RFC. My personal feeling it should be a layer above the TVM stack (similiar tvmc) to collect the artifacts produced out of tvm to generate such projects rather than making tvm a project generator. However, I dont have a strong opinion on either.

Thoughts ? : @tqchen

areusch replied:
I agree that we should discuss with an RFC. however, I don't want to break support for GDB debugging with source now.

regarding the target_kind attribute--we do need to be a bit careful with those, because they serve as keys to the autotuning logs. I don't know that a debug source directory should be part of identifying an autotuning result.

it seems like perhaps `export_library` could just allow you to specify the tempdir as a kwarg?

manupa-arm replied:
I added another optional argument for export_library as workspace_dir. Technically, I could've use kwarg but I think that is for the purpose of pass through arguments to fcompile function. Moreover, I think have a workspace dir should be useful for debugging anyway for different use cases of export_library. See if this work :) .

manupa-arm replied:
https://github.com/apache/tvm/pull/7002/ -- Implemented here.

manupa-arm commented on include/tvm/ir/module.h:
```diff
@@ -56,11 +56,14 @@ class IRModuleNode : public Object {
   Map<GlobalTypeVar, TypeData> type_definitions;
   /*! \brief The source map for the module. */
   parser::SourceMap source_map;
+  /*! \brief The names of ext. functions for func registry */
```
Done

manupa-arm commented on src/relay/backend/contrib/codegen_c/codegen.cc:
```diff
@@ -215,28 +214,37 @@ class CodegenC : public MemoizedExprTranslator<std::vector<Output>>, public Code
-    code_stream_ << "#include <tvm/runtime/packed_func.h>\n";
-    code_stream_ << "#include <dlpack/dlpack.h>\n";
-    code_stream_ << "using namespace tvm::runtime;\n";
+    code_stream_ << "#include \"tvm/runtime/c_runtime_api.h\"\n";
```
Done

zhiics commented on include/tvm/ir/module.h:
```diff
@@ -56,11 +56,14 @@ class IRModuleNode : public Object {
   /*! \brief The source map for the module. */
   parser::SourceMap source_map;
+  /*! \brief The names of external functions for func registry */
+  Array<String> ext_func_names;
```
why is this needed? I think external functions should be able to get from the IRModule as well.

manupa-arm replied:
IIUC, by the time IRModule reaches codegen_c_host (See line 311), all the functions are PrimFunc s. Thus, it will not include the relay external functions that were taken out early in the compilation flow. However, when creating the system-lib those function names should be included in the func registry as they are also sort of "primitive" but externally compiled.

zhiics replied:
yeah, my point was that we probably don't need to have this in Relay.

areusch replied:
@zhiics I think i see what's happening here: with the c++ runtime on full OS, codegen could generate symbols from a variety of modules and rely on runtime lookup via dlsym. with C runtime which has no dlsym, we use a FuncRegistry which is currently generated inside the host module. this has led to the need for "model-level" metadata being passed to the codegen in a couple of cases. probably the right thing to do is move FuncRegistry to its own module and pass this metadata there?

zhiics replied:
ahh, I see. Then we don't really need to have changes in Relay if we move FuncRegistry to its own module and pass the functions names there, right?

areusch replied:
I think that is the right way to go...I don't have bandwidth for that til after TVM conf though :/

zhiics replied:
I would suggest we wait a bit to get on the right direction. @manupa-arm how do you think?

manupa-arm replied:
Would it be possible to mark this as a TODO (if that's not breaking or conflicting anything) and move on? IIUC, the suggestion is we would create another module that ends up being a source with func registry on it.

manupa-arm replied:
OK then, how about we use the MetaData module wrapper to capture the function names (inc. external ones -- as it has access to all the modules) and emit to devc.cc / devc.o ?

areusch replied:
I think the issue with metadata module is that it assumes it is just generating the data for storage rather than immediate consumption on the target device. it requires a load process that ultimately copies the data into RAM. we want to avoid this at all cost on µTVM, because RAM is most precious. so I think we either would need to modify metadatamodule to enforce alignment or create something else.

manupa-arm replied:
That is true for Packing of Imports which get unpacked in the runtime to recreate the runtime modules. I guess the question is would that be the only purpose of the metadata module ? since it is named as metadata module. Thus, here I am not suggesting not to pack, instead make the metadata module generate the func registry in the artifact (which is devc.cc/o) that it uses to pack the imports.

manupa-arm replied:
So this is my understanding, metadata module will pack non-DSO exportable imported modules and their constants to devc.{object_format) if they implement "get_symbol" and "get_const_vars", otherwise its virtually doing nothing (Correct me if I am wrong, @zhiics). In this case, when we want to create external c-source modules from our external codegen, metadata module's load process should be virtually harmless -- as the SavetoBinary will almost write nothing to devc.{object_format} file (but there are things being written)

So my question is since metadata module holds all the modules (DSO and non-DSO exportable which are just llvm/c modules) can generate the func_registry as well, if we want another entity to deal with model-level metadata. Do we really need another module for this ? I would really appreciate a way forward on this.  also : @tqchen

zhiics replied:
Sorry for the late response since it was a holiday. you are right, metadata module has the symbols and constants. BTW, are the symbols enough for you? It looks that you just need them to create the declarations.

manupa-arm replied:
Hi @zhiics , no worries and hope you had a good holiday :) 

I actually went down on a path to create a CSourceMetadata module that could encapsulate c-source modules and which in-turn could get encapsulated again by the current metadata module if some of the external function require constant offloading -- I think that could be decided based on the fact whether the external module implements "get_const_vars" and that has non-zero constants. I think if dont have constants that the external module requires to offload metadata module, we dont need to create the current metadata module.

Summarizing, Im thinking of having two layers of metadata modules:

dsoModule1      dsoModule2
|                              |
|                              |
 ----         |          -----
               |
 C-Source metadata module                      non-dso module (e.g., acl)
               |                                                    (that needs constant offloading)
               |                                                                          |
                 ------------------------| ------------------------------------
                                              |
                                    Metadata module

Thus, the first layer can hold the func registry (currently its implemented in both codegen_c_host and codegen_llvm). I think it could be just live in the C-source metadata module that can get compiled along with other DSOExportable modules. **I will put and RFC with a different PR to implement this change**. This way we also get to use the init process to do certain initialization for the BYOC external device.

manupa-arm replied:
Sorry :p tried my best drawing with text, but I think you should get what I mean.

zhiics replied:
@manupa-arm Thanks for more explanations. It looks good to me. This is exactly how metadata module has been used.

areusch replied:
@manupa-arm that also seems reasonable to me. one question though--do you still intend to keep FuncRegistry codegen as-is (i.e. would you still generate the same global symbols)? it would be nice to avoid worrying about alignment issues for global symbols if we are generating C.

manupa-arm replied:
No I've already to removed the FuncRegistry from codegen_host_c and codegen_llvm and make the runtime modules expose a packed func named : "get_func_names"  -- we can discuss the best name for it , I am not good at naming stuff :p . So the CSourceMetadata module will use from the runtime modules and create a single func registry as a seperate source to be linked together. Still fixing some tests, I ll put it up soon.

areusch replied:
ok, looking forward to it. sorry what I mean is: do you intend to output the same thing, even if it's just from a different module?

enumeration of the functions a TVMModule contains is a standing problem with the PackedFunc module infra. cc @tqchen @jroesch

manupa-arm replied:
Sorry, I am not sure I follow. Here what I meant is we could create the function registry in a different source that captures the function names from all the runtime modules and the packed func is also used only in the compile time to create c-source metadata module. That would be exactly same as what you would've produced inside the main "c"/"llvm" module except its created in a different module that has access to all the runtime modules created.

areusch replied:
cool--that makes sense to me.

manupa-arm replied:
https://github.com/apache/tvm/pull/7002/ -- Implemented here.

manupa-arm replied:
This is merged; hence resolving the conversation

manupa-arm commented:
Hi @areusch @zhiics ,

Happy new year to you all! .

After CSourceMetadata module, this PR got reduced to introducing a simple test for utvm byoc example and some adjustments to external c-source code generator to make it applicable for uTVM as well. We would like to carry this test to ensure external module compilation works in uTVM. 

Please take a look when you have time.

zhiics commented on src/relay/backend/contrib/codegen_c/codegen.cc:
```diff
@@ -229,25 +228,31 @@ class CSourceCodegen : public CSourceModuleCodegenBase {
-    code_stream_ << "#include <dlpack/dlpack.h>\n";
-    code_stream_ << "using namespace tvm::runtime;\n";
+    code_stream_ << "#include <tvm/runtime/c_backend_api.h>\n";
+    code_stream_ << "#include <tvm/runtime/crt/module.h>\n";
```
I think crt probably should be included only when needed (i.e. other use case don't really need it)

manupa-arm replied:
Done

zhiics commented on src/relay/backend/contrib/codegen_c/codegen.cc:
```diff
@@ -229,25 +228,31 @@ class CSourceCodegen : public CSourceModuleCodegenBase {
+      // These are only needed to handle metadata copying
+      code_stream_ << "#include <tvm/runtime/container.h>\n";
+      code_stream_ << "#include <tvm/runtime/packed_func.h>\n";
+      code_stream_ << "#include <dlpack/dlpack.h>\n";
```
I think packed_func is probably needed to handle packed args. dlpack has to be included if we have DLTensor. The reason it doesn't cause errors is probably because c_runtime_api.h includes it as well.

manupa-arm replied:
Yes, dlpack is in the c_runtime_api.h, thus we can remove it altogether.

Do we need packed_func.h ? -- I would like to avoid if the constants copying process is not required. The obvious usage is from TVM_DLL_EXPORT_TYPED_FUNC macro here which creates C++ code. Hence the gating based on presence of constants.

zhiics replied:
@areusch Yeah, we used it to generated C++ code before when working on BYOC. But I think you are right, we should just export C code and get rid of the dependency of TVM_DLL_EXPORT_TYPED_FUNC

manupa-arm replied:
I removed dlpack for now and since we need to have Array<NDArray> here, we cant get away not using C++ here. Therefore, I might as well use the macro for convenience here.

However, I made it obvious that this part is not codegen'd/ not visible for a c-compiler.

areusch commented on tests/micro/qemu/test_zephyr.py:
```diff
@@ -198,5 +200,143 @@ def test_relay(platform):
+    with tvm.transform.PassContext(opt_level=3, config={"tir.disable_vectorize": True}):
+        graph, mod, params = tvm.relay.build(relay_mod, target=target)
+
+    with _make_session(model, target, zephyr_board, mod) as session:
```
I think technically right now this invokes the C++ compiler (due to limitation in `export_library`). at some point in the future, i think the true test you want would be one which invokes the C compiler--does that match your understanding?

manupa-arm replied:
Well, I changed the export library to write .c files now in #7002 If you can re-call.

manupa-arm replied:
So it should invoke c compiler now, at least when we wrap up the fcompile with a micro.library instance.

areusch replied:
ah that's correct, thanks for jogging my memory. great, so this test should do everything you need for now.

areusch reviewed (changes requested):
@manupa-arm happy new year and thanks for continuing to push this forward. this looks pretty good, just a couple small things and some clarifying questions (as I'm not as familiar with codegen_c internals as I should be).

manupa-arm commented on tests/micro/qemu/test_zephyr.py:
```diff
@@ -198,5 +200,143 @@ def test_relay(platform):
         tvm.testing.assert_allclose(result, x_in * x_in + 1)
 
 
+class CcompilerAnnotator(ExprMutator):
```
I think running on the bare metal is the requirement and thus would prefer to keep it here and supported with future bare metal related developements.

areusch replied:
okay and actually since this involves a separate codegen path, it makes sense to keep this here.

manupa-arm commented on src/relay/backend/contrib/codegen_c/codegen.cc:
```diff
@@ -229,25 +228,31 @@ class CSourceCodegen : public CSourceModuleCodegenBase {
+      code_stream_ << "#include <tvm/runtime/container.h>\n";
+      code_stream_ << "#include <tvm/runtime/packed_func.h>\n";
+      code_stream_ << "#include <dlpack/dlpack.h>\n";
+      code_stream_ << "using namespace tvm::runtime;\n";
```
Actually this part is of segment is only to be generated in the presence of constant copying process and makes this source a C++ one. To this end, there is a init function that accepts "Array<NDArray> arr" (See in codegen_c.h) that is in that namespace. I could not think of a way to make it a C source immediately.

Honestly, I was less worried about it as this is an example codegen and in utvm we dont want to be doing constant copying anyway. Therefore this segment will not be codegen'd.

areusch replied:
hrm. should we then add code like:

```
#ifdef __cplusplus
using std::malloc;
#endif
```

you're right this is a demo codegen, perhaps we don't need to resolve this. it's just confusing to create c++ code in a codegen named codegen_c.

manupa-arm replied:
@areusch So I made it obvious this segment is C++ in the code. See if it works.

manupa-arm commented on tests/micro/qemu/test_zephyr.py:
```diff
@@ -198,5 +200,143 @@ def test_relay(platform):
+
+def check_result(relay_mod, model, zephyr_board, map_inputs, out_shape, result):
+    """Helper function to verify results"""
+    tol = 1e-5
```
Done

manupa-arm commented on src/relay/backend/contrib/codegen_c/codegen.cc:
```diff
@@ -229,25 +228,31 @@ class CSourceCodegen : public CSourceModuleCodegenBase {
+    code_stream_ << "#include <tvm/runtime/c_backend_api.h>\n";
+    code_stream_ << "#include <tvm/runtime/crt/module.h>\n";
+    if (!variables.empty()) {
+      // These are only needed to handle metadata copying
```
Done and added a longer description to say why C++ code is generated.

manupa-arm commented:
Hi @areusch @zhiics ,

I've addressed all the minor comments.

The current limitation of making this generator a pure C code generator is that is carries a Array\<NDArray\> for constants that will be copied in, if required by the compilation pipeline. I feel changing this is a bigger work and its only there to test the external codegen/runtime pathway (especially the JSONRuntime) handles the constant copying. Therefore, given that we  have Array\<NDArray\>, I might as well use the macro. However, I clearly commented and made it obvious that bit of the code is C++ (using ifdef __cplusplus as @areusch suggested.) 

Therefore, in the absense of the constant copying, this will generate a pure C code -- If you see I have removed macro usage for the main compute function (its just the constant copying one uses it).

Let me know your thoughts and what more could be done about it, if any.

areusch reviewed (approved):
looks good, thanks for bearing with me @manupa-arm!

zhiics reviewed (approved):
LGTM. Thanks for the effort.

zhiics commented:
Thanks @manupa-arm @areusch
